### ISSUE TITLE
DataTable Toolbar Search not working

### ISSUE THREAD

**Issue by pkruszynski** [carbon-design-system/carbon-components-svelte#591]
I'm writing Svelte app that requires table row filtering and Carbon `DataTable` with `ToolbarSearch` looked very promising.

```svelte
<script>
  import {
    DataTable,
    Toolbar,
    ToolbarContent,
    ToolbarSearch
  } from "carbon-components-svelte";

  const headers = [
    { key: "name", value: "Name" },
    { key: "port", value: "Port" },
    { key: "rule", value: "Rule" },
  ];

  const rows = [
    { id: "a", name: "Load Balancer 3", port: 3000, rule: "Round robin" },
    { id: "b", name: "Load Balancer 1", port: 443, rule: "Round robin" },
    { id: "c", name: "Load Balancer 2", port: 80, rule: "DNS delegation" },
    { id: "d", name: "Load Balancer 6", port: 3000, rule: "Round robin" },
    { id: "e", name: "Load Balancer 4", port: 443, rule: "Round robin" },
    { id: "f", name: "Load Balancer 5", port: 80, rule: "DNS delegation" }
  ];
</script>

<DataTable {headers} {rows} >
  <Toolbar>
    <ToolbarContent>
      <ToolbarSearch />
    </ToolbarContent>
  </Toolbar>
</DataTable>
```

Unfortunately `ToolbarSearch` is not working - is this broken or not implemented? And if not implemented when is it coming?

**Comment by metonym** [comment]
Filtering behavior is not currently supported. When the toolbar search was implemented, this was an intentional design choice. However, I think I've seen a similar question raised by someone else, perhaps there is a need for it.

> IMO, any row filtering logic should be deferred to the consumer.

https://github.com/IBM/carbon-components-svelte/issues/14#issue-539624322

In the meantime, you can implement basic filtering like so:

```svelte
<script>
  import {
    DataTable,
    Toolbar,
    ToolbarContent,
    ToolbarSearch
  } from "carbon-components-svelte";

  const headers = [
    { key: "name", value: "Name" },
    { key: "port", value: "Port" },
    { key: "rule", value: "Rule" },
  ];

  const rows = [
    { id: "a", name: "Load Balancer 3", port: 3000, rule: "Round robin" },
    { id: "b", name: "Load Balancer 1", port: 443, rule: "Round robin" },
    { id: "c", name: "Load Balancer 2", port: 80, rule: "DNS delegation" },
    { id: "d", name: "Load Balancer 6", port: 3000, rule: "Round robin" },
    { id: "e", name: "Load Balancer 4", port: 443, rule: "Round robin" },
    { id: "f", name: "Load Balancer 5", port: 80, rule: "DNS delegation" }
  ];

  let value = '';

  $: filteredRows = rows
        .filter((row) => {
          if (value.trim().length === 0) return row;
          return row.name.includes(value);
        });
</script>

<DataTable {headers} rows={filteredRows} >
  <Toolbar>
    <ToolbarContent>
      <ToolbarSearch bind:value />
    </ToolbarContent>
  </Toolbar>
</DataTable>
```

**Comment by pkruszynski** [comment]
Thank you for prompt and detailed answer @metonym - this is great!
In the meantime, expanding on your example, I implemented case insensitive filtering for all example table columns (apart from `id` column):
```svelte
<script>
  import {
    DataTable,
    Toolbar,
    ToolbarContent,
    ToolbarSearch
  } from "carbon-components-svelte";

  const headers = [
    { key: "name", value: "Name" },
    { key: "port", value: "Port" },
    { key: "rule", value: "Rule" },
  ];

  const rows = [
    { id: "a", name: "Load Balancer 3", port: 3000, rule: "Round robin" },
    { id: "b", name: "Load Balancer 1", port: 443, rule: "Round robin" },
    { id: "c", name: "Load Balancer 2", port: 80, rule: "DNS delegation" },
    { id: "d", name: "Load Balancer 6", port: 3000, rule: "Round robin" },
    { id: "e", name: "Load Balancer 4", port: 443, rule: "Round robin" },
    { id: "f", name: "Load Balancer 5", port: 80, rule: "DNS delegation" }
  ];

  let value = '';

  $: filteredRows = rows
        .filter((row) => {
          if (value.trim().length === 0) return row;
          return (row.name.toLowerCase().includes(value) || row.port.toString().includes(value) || row.rule.toLowerCase().includes(value));
        });
</script>

<DataTable {headers} rows={filteredRows} >
  <Toolbar>
    <ToolbarContent>
      <ToolbarSearch bind:value />
    </ToolbarContent>
  </Toolbar>
</DataTable>
```

**Comment by metonym** [comment]
Auto-filtering would work especially well with paginated datatables where the consumer provides all the rows.

**Comment by metonym** [comment]
This feature makes a lot of sense now that paginated data tables are supported (https://github.com/carbon-design-system/carbon-components-svelte/pull/880).

It would address the use case of filtering a table in-memory on the client side, say with <500 rows.

`DataTable` could use Svelte Context to pass `rows` to `ToolbarSearch`.

**Comment by metonym** [comment]
This is now supported in Released in [v0.62.0](https://github.com/carbon-design-system/carbon-components-svelte/releases/tag/v0.62.0).

[**Auto-filtering**](https://carbon-components-svelte.onrender.com/components/DataTable#filterable)

Set `shouldFilterRows` to true to enable auto-filtering. It performs a basic string comparison against row cell values that are a string or number type.

```svelte
<ToolbarSearch shouldFilterRows />
```

[**Custom filtering**](https://carbon-components-svelte.onrender.com/components/DataTable#filterable-custom)

For custom filtering, `shouldFilterRows` also accepts a function and passes it the current row and value. It expects the function to return a boolean.

```svelte
<ToolbarSearch
  shouldFilterRows={(row, value) => {
    return (
      /(6|8)$/.test(row.name) &&
      row.rule.toLowerCase().includes(value.toLowerCase())
    );
  }}
/>
```
